### PR TITLE
Format field in KubeadmConfig.Spec

### LINKED ISSUE #121: Add a field to the spec called Format
/kind feature

**Describe the solution you'd like**
Following up from a discussion https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/issues/11#issuecomment-517698950

We are going to support multiple output versions *after* v1a2 but we are going to add the field now so that we don't have to modify the API later.

Don't forget to regenerate everything after adding the field!

/good-first-issue
/priority important-soon
/milestone v0.1

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:
This PR adds the field `Format` to `KubeadmConfig.Spec`. 
The `Format` is a required field that specifies the output format of the bootstrap data.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #121 

**Special notes for your reviewer**:
The `Format` field needs to be wired up. An issue will be opened to discuss how to use the field in the controller.

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

File: api/v1alpha2/kubeadmbootstrapconfig_types.go [hunk 1]
```diff
@@ -21,6 +21,12 @@ import (
 	kubeadmv1beta1 "sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm/kubeadm/v1beta1"
 )
 
+type Format string
```
Comment (vincepri) [inline api/v1alpha2/kubeadmbootstrapconfig_types.go]: We should add something like
`// +kubebuilder:validation:Enum=cloud-config`

Comment (ykakarap) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: Done :)

**Comment by ykakarap** [comment]
I want to confirm that this field is not an optional field. It is a mandatory field.

**Comment by detiber** [comment]
> For the sake of UX simplicity, IMO this flag should be optional/defaulted to cloud-config

Since we don't have the mutating webhook in place, we don't really have a good way to default today outside of the implementation. +1 to making this field optional, though.

File: api/v1alpha2/kubeadmbootstrapconfig_types.go (hunk 1 above)
Comment (chuckha) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: nice

**Comment by ykakarap** [comment]
@fabriziopandini @detiber I made the field optional.
As @detiber there is currently no mechanism to set a default. We have to make sure to set a default in the implementation while using the field.

**Comment by detiber** [comment]
/lgtm
/hold
holding for further review by @fabriziopandini and @vincepri

**Comment by chuckha** [comment]
lgtm pending the `make generate-manifests`

**Comment by ykakarap** [comment]
>lgtm pending the make generate-manifests

I already did that. The YAML is already updated with the correct information. I wonder why it failed 🤔

Re-running tests.

/retest

**Comment by chuckha** [comment]
@amy Any ideas on this test failure?

**Comment by chuckha** [comment]
Discussed offline. This confusing error is because prow automatically rebases the patch.

**Comment by chuckha** [comment]
hold for commit squashing

**Pull request merged**